rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java

```diff
@@ -2002,5 +2005,97 @@ private void ensureBtcBlockStore() throws IOException, BlockStoreException {
+            } else {
+                logger.warn("whitelist money return tx build for btc tx {} error. Return was to {}. Tx {}. Value {}", btcTx.getHash(), senderBtcAddress, rskTx, totalAmount);
+                panicProcessor.panic("whitelist-return-funds", String.format("whitelist money return tx build for btc tx %s error. Return was to %s. Tx %s. Value %s", btcTx.getHash(), senderBtcAddress, rskTx, totalAmount));
+            }
+            return false;
+        }
+
+        // Consume this whitelisted address
+        lockWhitelist.consume(senderBtcAddress);
+
+        return true;
+    }
+
+    private boolean verifyLockDoesNotSurpassLockingCap(Transaction rskTx, BtcTransaction btcTx, Address senderBtcAddress, Coin totalAmount) throws IOException {
+        if (!activations.isActive(ConsensusRule.RSKIP134)) {
+            return true;
+        }
+
+        Coin fedCurrentFunds = getFederationCurrentFunds();
+        Coin lockingCap = this.getLockingCap();
+        logger.trace("Evaluating locking cap for: Sender {}. Value to lock {}. Current funds {}. Current locking cap {}", senderBtcAddress, totalAmount, fedCurrentFunds, lockingCap);
+        Coin fedUTXOsAfterThisLock = fedCurrentFunds.add(totalAmount);
+        // If the federation funds (including this new UTXO) are smaller than or equals to the current locking cap, we are fine.
+        if (fedUTXOsAfterThisLock.compareTo(lockingCap) <= 0) {
+            return true;
+        }
+
+        logger.info("locking cap exceeded! btc Tx {}", btcTx);
+        // Reject the lock
+        Optional<ReleaseTransactionBuilder.BuildResult> buildReturnResult = this.getRefundingTransaction(btcTx, senderBtcAddress);
+        if (buildReturnResult.isPresent()) {
+            // Send the release transaction to request signature immediately
+            provider.getRskTxsWaitingForSignatures().put(rskTx.getHash(), buildReturnResult.get().getBtcTx());
+            logger.info("money return tx build successful to {}. Tx {}. Value {}.", senderBtcAddress, rskTx, totalAmount);
+        } else {
+            logger.warn("money return tx build for btc tx {} error. Return was to {}. Tx {}. Value {}", btcTx.getHash(), senderBtcAddress, rskTx, totalAmount);
+            panicProcessor.panic("locking-cap-exceeded-return-funds", String.format("locking cap exceeded! money return tx build for btc tx %s error. Return was to %s. Tx %s. Value %s", btcTx.getHash(), senderBtcAddress, rskTx, totalAmount));
+        }
+        return false;
+    }
+
+    private Optional<ReleaseTransactionBuilder.BuildResult> getRefundingTransaction(BtcTransaction btcTx, Address senderBtcAddress) throws IOException {
+
+        // Build the list of UTXOs in the BTC transaction sent to either the active
+        // or retiring federation
+        List<UTXO> utxosToUs = btcTx.getWalletOutputs(getNoSpendWalletForLiveFederations()).stream()
```

Comment:
This is not our fault, but "getWalletOutputs" is a completely misleading name. It should be called "getOutputs" or "getTransactionOutputs". "Wallet" has nothing to do. I would prefer we don't use this method and we wrap it with an utility function getTransactionOutputs(btcTx,..) that uses getWalletOutputs with a big comment about this.